File: 0064-minimum-path-sum/0064-minimum-path-sum.cpp

#define vi vector<int>
#define vvi vector<vi>

class Solution {
public:

    int recurSol(vvi& grid, int n, int m) {
    
        if(n==0 && m==0) return grid[n][m];

        int x = INT_MAX;
        int y = INT_MAX;

        if(n-1>=0)
            x = grid[n][m] + recurSol(grid,n-1,m);
        if(m-1>=0)
            y = grid[n][m] + recurSol(grid,n,m-1);

        return min(x,y);

    }

    int memoSol(vvi& grid, int n, int m,vvi& dp) {
    
        if(n==0 && m==0) return dp[n][m] = grid[n][m];

        int x = INT_MAX;
        int y = INT_MAX;

        if(dp[n][m]!=-1) return dp[n][m];

        if(n-1>=0)
            x = grid[n][m] + memoSol(grid,n-1,m,dp);
        if(m-1>=0)
            y = grid[n][m] + memoSol(grid,n,m-1,dp);

        return dp[n][m] = min(x,y);

    }
    int minPathSum(vector<vector<int>>& grid) {

        int n = grid.size();
        int m = grid[0].size();

        // return recurSol(grid, n-1, m-1);
        vvi dp(n,vi(m,-1));
        return memoSol(grid,n-1,m-1,dp);
        
    }
};

File: 0064-minimum-path-sum/0064-minimum-path-sum.java
class Solution {
    public int solve(int[][] grid, int x, int y) {

        if(x==0 && y==0) {
            return grid[x][y];
        }

        if(x<0 || y<0) return Integer.MAX_VALUE;

        int up = solve(grid,x-1,y);
        int left = solve(grid,x,y-1);

        return grid[x][y]+Math.min(up,left);
    }

    public int solve2(int[][] grid, int x, int y,int[][] dp) {

        if(x==0 && y==0) {
            return grid[x][y];
        }

        if(x<0 || y<0) return Integer.MAX_VALUE;

        if(dp[x][y]!=0) return dp[x][y];

        int up = solve2(grid,x-1,y,dp);
        int left = solve2(grid,x,y-1,dp);

        return dp[x][y] = grid[x][y]+Math.min(up,left);
    }

    public int solve3(int[][] grid) {
        int n  = grid.length;
        int m = grid[0].length;

        int[][] dp = new int[n][m];

        dp[0][0] = grid[0][0];

        for(int i = 0;i<n;i++) {
            for(int j = 0;j<m;j++) {
                if(i==0 && j==0) continue;

                int left = (j-1<0)?Integer.MAX_VALUE:dp[i][j-1];
                int up = (i-1<0)?Integer.MAX_VALUE:dp[i-1][j];

                dp[i][j] = grid[i][j]+Math.min(left,up);

            }
        }

        return dp[n-1][m-1];

    }

    public int minPathSum(int[][] grid) {
        int n = grid.length;
        int m = grid[0].length;
        int x = n-1;
        int y = m-1;

        int ans = 0;
        // ans = solve(grid,x,y);

        // int[][] dp = new int[n][m];

        // ans = solve2(grid,x,y,dp);
        ans = solve3(grid);
        return ans;
    }
}